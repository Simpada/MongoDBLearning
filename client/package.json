{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "build": "npm install --include=dev && parcel build index.html",
    "dev": "parcel watch index.html",
    "test": "jest"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@testing-library/react": "^13.4.0",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "parcel": "^2.8.0",
    "process": "^0.11.10"
  },
  "dependencies": {
    "@parcel/transformer-react-refresh-wrap": "^2.8.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.3"
  },
  "babel": {
    "presets": [
      "@babel/preset-react",
      [
        "@babel/preset-env",
        {
          "targets": {
            "node": "current"
          }
        }
      ]
    ]
  },
  "jest": {
    "testEnvironment": "jsdom",
    "collectCoverageFrom": [
      "**/*.jsx",
      "**/*.js",
      "!dist/**",
      "!coverage/**"
    ]
  }
}
