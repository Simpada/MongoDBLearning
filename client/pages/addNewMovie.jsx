import {useState} from "react";
import * as React from "react";
import {fetchJSON} from "../index";


export function AddNewMovie() {

    const [title, setTitle] = useState("");
    const [year, setYear] = useState("");
    const [plot, setPlot] = useState("");
    const [genres, setGenres] = useState("");

    async function handleSubmit(e) {
        e.preventDefault();

        await fetchJSON("/api/movies", {
            method: "post",
            json: { title, year: parseInt(year), plot, genres },
        });

        setTitle("");
        setYear("");
        setPlot("");
        setGenres("");
    }

    return (
        <form onSubmit={handleSubmit}>
            <h1>Add new movie</h1>
            <div>
                Title:
                <input value={title} onChange={(e) => setTitle(e.target.value)} />
            </div>
            <div>
                Year:
                <input value={year} onChange={(e) => setYear(e.target.value)} />
            </div>
            <div>
                Plot:
                <input value={plot} onChange={(e) => setPlot(e.target.value)} />
            </div>
            <div>
                Genre:
                <input value={genres} onChange={(e) => setGenres(e.target.value)} />
            </div>
            <div>
                <button disabled={title.length === 0 || year.length === 0}>Save</button>
            </div>
        </form>
    );
}