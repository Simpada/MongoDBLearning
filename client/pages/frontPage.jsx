import {Link} from "react-router-dom";
import * as React from "react";

export function FrontPage() {
    return <div>
        <h1>Movie Database</h1>
        <ul>
            <li><Link to={"/movies"}>List Movies</Link></li>
            <li><Link to={"/movies/new"}>Add New Movie</Link></li>
        </ul>
    </div>
}