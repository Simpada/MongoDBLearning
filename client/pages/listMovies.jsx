import * as React from "react";
import {useLoading} from "../useLoading";


function MovieCard({movie: {title, poster, plot, year, genres}}) {
    return <><h3>{title}</h3>
        {poster && <img src={poster} alt={"Movie Poster"} width={100}/>}
        <div>{plot}</div>
        <div>{year}</div>
        <div>{genres}</div>
    </>;
}

export function ListMovies({listMovies}) {

    const {loading, error, data} = useLoading(listMovies
    );

    if (loading) {
        return <div>Loading...</div>
    }
    if (error) {
        return <div>
            <h1>ERROR</h1>
            <div>{error.toString()}</div>
        </div>
    }

    return <div>
        <h1>Movies in the database</h1>

        {data.map((movie) => (
            <MovieCard key={movie.title} movie={movie}/>
        ))}

    </div>;
}