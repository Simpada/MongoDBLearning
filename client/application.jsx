import {BrowserRouter, Route, Routes} from "react-router-dom";
import {FrontPage} from "./pages/frontPage";
import {ListMovies} from "./pages/listMovies";
import {AddNewMovie} from "./pages/addNewMovie";
import React from "react";
import {fetchJSON} from "./index";

export function Application() {

    async function listMovies() {
        return await fetchJSON("/api/movies")
    }

    return <BrowserRouter>
        <Routes>
            <Route path={"/"} element={<FrontPage/>}/>
            <Route path={"/movies"} element={<ListMovies listMovies={listMovies}/>}/>
            <Route path={"/movies/new"} element={<AddNewMovie/>}/>
        </Routes>
    </BrowserRouter>
        ;
}