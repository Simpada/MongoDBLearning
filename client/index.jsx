import * as React from "react";
import {createRoot} from "react-dom/client";
import {Application} from "./application";

const element = document.getElementById("app");
const root = createRoot(element);


export async function fetchJSON(url) {
    const res = await fetch(url);
    if (!res.ok) {
        throw new Error(`Failed ${res.status}: ${(await res).statusText}`);
    }
    return await res.json();
}

root.render(<Application/>);
