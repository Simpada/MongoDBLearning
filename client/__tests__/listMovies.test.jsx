import {createRoot} from "react-dom/client";
import * as React from "react";
import {ListMovies} from "../pages/listMovies";
import {act} from "react-dom/test-utils";

describe("ListMovies component", () => {
    it("shows loading screen", () => {

        const domElement = document.createElement("div");
        const root = createRoot(domElement);
        root.render(<ListMovies/>);

        expect(domElement.innerHTML).toMatchSnapshot();
    });
    it("shows movie", async () => {

        const movies = [{title: "movie 1"}, {title: "movie 2"}];
        const domElement = document.createElement("div");
        await act(async () => {
            const root = createRoot(domElement);
            root.render(<ListMovies listMovies={() => movies}/>);
        });

        expect(Array.from(domElement.querySelectorAll("h3"))
            .map(e => e.innerHTML))
            .toEqual(["movie 1", "movie 2"]);
        expect(domElement.innerHTML).toMatchSnapshot();
    });
})