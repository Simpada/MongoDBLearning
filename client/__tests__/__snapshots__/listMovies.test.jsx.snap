// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovies component shows error message 1`] = `"<div><h1>ERROR</h1><div id="error-text">Error: Something went wrong!</div></div>"`;

exports[`ListMovies component shows loading screen 1`] = `""`;

exports[`ListMovies component shows movie 1`] = `"<div><h1>Movies in the database</h1><h3>movie 1</h3><div></div><div></div><div></div><h3>movie 2</h3><div></div><div></div><div></div></div>"`;
